src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp

```diff
@@ -2476,6 +2480,287 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+        __ cmpq(rScratch2, 32);
+        __ jccb(Assembler::below, L_dwordsTail);
+        __ leaq(rScratch3, Address(dest, 28));
+        __ movq(rScratch4, rScratch1);
+
+        __ BIND(L_dwordLoop);
+
+        // Unroll 8 dword stores
+        for (int i = 7; i >= 0; i--) {
+          __ movl(Address(rScratch3, -(i * 4)), wide_value);
+        }
+        __ addq(rScratch3, 32);
+        __ decrementq(rScratch4);
+        __ jccb(Assembler::notZero, L_dwordLoop);
+
+        __ BIND(L_dwordsTail);
+
+        // Handle leftovers
+        __ shlq(rScratch1, 3);
+        __ shrq(rScratch2, 2);
+        __ cmpq(rScratch1, rScratch2);
+        __ jccb(Assembler::aboveEqual, L_exit);
+        __ decrementq(size);
+        __ shrq(size, 2);
+        __ incrementq(size);
+
+        __ BIND(L_dwordsTailLoop);
+
+        __ movl(Address(dest, rScratch1, Address::times_4), wide_value);
+        __ incrementq(rScratch1);
+        __ cmpq(size, rScratch1);
+        __ jccb(Assembler::notEqual, L_dwordsTailLoop);
+      }
+      __ jmp(L_exit);
+
+      __ BIND(L_fillBytes);
+    }
+#ifdef MUSL_LIBC
+    {
+      Label L_byteLoop, L_longByteLoop, L_byteTail, L_byteTailLoop;
+
+      const Register savedSize = rax;
+      const Register byteVal = rdx;
+
+      UnsafeSetMemoryMark usmm(this, true, true);
+
+      __ movq(savedSize, size);
+      __ andq(savedSize, 7);
+      __ cmpq(size, 8);
+      __ jccb(Assembler::aboveEqual, L_byteLoop);
+      __ xorl(rScratch1, rScratch1);
+      __ jmpb(L_byteTail);
+
+      __ BIND(L_byteLoop);
+
+      __ andq(size, -8);
+      __ xorl(rScratch1, rScratch1);
+
+      __ BIND(L_longByteLoop);
+
+      // Unroll 8 byte stores
+      for (int i = 0; i < 8; i++) {
+        __ movb(Address(dest, rScratch1, Address::times_1, i), byteVal);
+      }
+
+      __ addq(rScratch1, 8);
+      __ cmpq(size, rScratch1);
+      __ jccb(Assembler::notEqual, L_longByteLoop);
+
+      __ BIND(L_byteTail);
+
+      __ testq(savedSize, savedSize);
+      __ jccb(Assembler::zero, L_exit);
+      __ addq(dest, rScratch1);
+      __ xorl(rScratch1, rScratch1);
+
+      __ BIND(L_byteTailLoop);
+
+      __ movb(Address(dest, rScratch1, Address::times_1), byteVal);
+      __ incrementq(rScratch1);
+      __ cmpq(savedSize, rScratch1);
+      __ jccb(Assembler::notEqual, L_byteTailLoop);
+    }
+    __ jmp(L_exit);
+#else  // MUSL_LIBC
+    {
+      const Register byteVal = rdx;
+#ifdef _WIN32
+      __ movq(rcx, rdi); // Restore c_rarg*
+      __ movq(r8, rdx);
+      __ movq(rdx, rsi);
+      restore_arg_regs();
+#endif
+      // rax has expanded byte value
+      __ movq(c_rarg2, rax);
+
+      __ xchgq(c_rarg1, c_rarg2);
+      // generate_unsafe_fill(T_BYTE, false, "unsafe_set_memory");
+      __ mov(r11, c_rarg2);
+
+      __ enter(); // required for proper stackwalking of RuntimeStub frame
+
+      {
+        UnsafeSetMemoryMark usmm(this, true, true);
+
+        __ generate_fill(T_BYTE, false, c_rarg0, c_rarg1, r11, rax, xmm0);
```

Comment:
We will be duplicating the code gen for generate_fill here? Could we not do a tail call to _jbyte_fill here and add UnsafeSetMemoryMark inside _jbyte_fill?

Reply:
It would not be appropriate to add set memory marks to the existing _jbyte_fill as it is being used by other routines, and the effect of the mark will be very hard to track down (if any).

Are you *sure* we want to do that?

Reply:
Yes we want to do that. It is all guarded by thread->doing_unsafe_access() which is only true when we are getting to this code from unsafe. Similar technique is used in copyMemory as well.

Reply:
Done.